core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -221,28 +222,172 @@ async def process_internal_input(self, txi: TxInput) -> None:
     async def process_external_input(self, txi: TxInput) -> None:
         self.approver.add_external_input(txi)
 
-    async def approve_output(self, txo: TxOutput, script_pubkey: bytes) -> None:
-        if self.output_is_change(txo):
-            # output is change and does not need approval
+    async def process_original_input(self, txi: TxInput) -> None:
+        assert txi.orig_hash is not None
+        assert txi.orig_index is not None
+
+        for orig in self.orig_txs:
+            if orig.orig_hash == txi.orig_hash:
+                break
+        else:
+            orig_meta = await helpers.request_tx_meta(
+                self.tx_req, self.coin, txi.orig_hash
+            )
+            orig = OriginalTxInfo(self, orig_meta, txi.orig_hash)
+            self.orig_txs.append(orig)
+
+        if txi.orig_index >= orig.tx.inputs_count:
+            raise wire.ProcessError("Not enough inputs in original transaction.")
+
+        if orig.index != txi.orig_index:
+            raise wire.ProcessError(
+                "Rearranging or removal of original inputs is not supported."
+            )
+
+        orig_txi = await helpers.request_tx_input(
```

Comment:
I had actually considered this, but it would be a little bit more complex:

> * `orig_txi` may include `script_sig` or `witness` -- but we can instead insist on the host providing these as part of `txi` whenever `orig_hash` is set

1. You always need the `script_sig` of the original input to compute the TXID correctly.
2. You need the `script_sig` of the replacement input if it's a pre-signed external input.
3. You sometimes need the original `witness` to verify the signature of an internal input in the original transaction.
4. You sometimes need the `witness` field in the replacement transaction for pre-signed external inputs.

So we would need to add an `orig_script_sig` field to be able to compute the TXIDs. Cases 3 and 4 are mutually exclusive, so one `witness` field would be sufficient, but maybe it would be worth adding `orig_witness` just for consistency and clarity (avoiding logic of the sort, "if it's external it means this, if it's internal it means that").
 
> * `orig_txi` may change `sequence` in order to disable RBF in the replacement

As you noted, we can disable this use-case and if we need it in the future, then we can add an `orig_sequence` field.

> the below code, which could otherwise be replaced by `orig.add_input(txi)` without re-streaming the same input

I would have to think about this again, but I think I had come to the conclusion that the code around `write_tx_input()` or `write_tx_input_check()` would start getting a bit messy. Either we'd have to add something like `write_tx_orig_input()`, or we would have to instantiate a `TxInput` and copy `orig_script_sig` into `script_sig` etc., so that when computing the various hashes you use the right `script_sig`, `sequence` etc. My memory is a bit fuzzy on this, but this is the general idea of what discouraged me.

To be consistent it would also then make sense to do the same for outputs, like adding the `orig_amount` field. Again same issues as above. (I realize there is a difference here, that outputs can be removed, whereas inputs cannot, but actually we could support removal of external inputs in the future, we just currently don't have a use-case to justify the added code complexity.)

So it would reduce the number of messages exchanged and eliminate the need for the check below, but then it seems it would add quite a bit more complexity elsewhere. Finally, the whole idea seems to against the recent changes to messages where we tried to simplify the messages by introducing a message for each purpose, like `TxInput` and `PrevInput` rather than packing one message type with lots of fields, which also makes the protocol logic more complex for the host.

Reply:
so out of `TxInput`, the following must always match between orig and current:
`prev_hash`, `prev_index`, `amount`, `script_type`

the following must always differ:
`script_sig + witness`
(for external inputs, the signature between original and replacement differs; for internals, the original has the fields while the current doesn't)

the following _may_ differ:
`sequence` (the only parameter where modification is meaningful)

we do not care about:
`address_n` (+ multisig) -- its _presence_ in the orig indicates verifiable input, but the actual value seems not to matter?
in which case we could also count it as "must always match"

what about `ownership_proof + commitment_data`? does the same ownership proof apply for both original and replacement? if yes, would it make sense to generate a _different_ one for the replacement?

this slightly suggests the possibility of `TxOrigInput` type which only has `script_sig, witness, sequence` (maybe ownership proof) -- but I suspect that would make the code more complex elsewhere, namely in `get_legacy_tx_digest()`

---

in conclusion, let's keep the streaming of previous inputs as-is, but please add a comment explaining why only three of the values are checked for matching (and consider adding `script_type` to the list, unless there's something i'm missing?)

---

also: perhaps a chunk of the `process_original_input()` should be moved to `TxOrigInfo.add_input`? I can see that it wouldn't truly be "info" anymore.

Reply:
> `address_n` (+ multisig) -- its _presence_ in the orig indicates verifiable input, but the actual value seems not to matter?

Correct.

> what about `ownership_proof + commitment_data`? does the same ownership proof apply for both original and replacement? if yes, would it make sense to generate a _different_ one for the replacement?

Cryptographically it is valid for both, but we are not interested in it for original inputs (at least until we implement support for removal of external inputs which, as I mentioned elsewhere, doesn't have a strong use-case at the moment).

> [...] but I suspect that would make the code more complex elsewhere, namely in `get_legacy_tx_digest()`

Exactly.

> also: perhaps a chunk of the `process_original_input()` should be moved to `TxOrigInfo.add_input`? I can see that it wouldn't truly be "info" anymore.

To me it doesn't belong in TxInfo, the signer should be responsible for driving the transaction flow, like calling `helpers.request_tx_input()`.

> in conclusion, let's keep the streaming of previous inputs as-is, but please add a comment explaining why only three of the values are checked for matching (and consider adding `script_type` to the list, unless there's something i'm missing?)

I added `script_type`, because it should always match, but from a security perspective it seems irrelevant, because we make the internal/external decision based on the replacement `script_type`.

See https://github.com/trezor/trezor-firmware/pull/1292/commits/c63f0bc8842c87d8c8a8e2142a4a8b919aaedf50